src/realm/object-store/sectioned_results.cpp

```diff
@@ -276,42 +255,57 @@ ResultsSection::ResultsSection(SectionedResults* parent, Mixed key)
 bool ResultsSection::is_valid() const
 {
     util::CheckedUniqueLock lock(m_parent->m_mutex);
+    if (!m_parent->is_valid())
+        return false;
+    // See if we need to recalculate the sections before
+    // searching for the key.
     m_parent->calculate_sections_if_required();
-    auto it = m_parent->m_sections.find(m_key);
-    return it != m_parent->m_sections.end() && m_parent->is_valid();
+    if (m_parent->m_sections.find(m_key) == m_parent->m_sections.end())
+        return false;
+
+    return true;
 }
 
 Mixed ResultsSection::operator[](size_t idx) const
 {
     util::CheckedUniqueLock lock(m_parent->m_mutex);
+    if (!m_parent->is_valid())
+        throw std::logic_error("The parent Results collection is not valid.");
     m_parent->calculate_sections_if_required();
     auto it = m_parent->m_sections.find(m_key);
     REALM_ASSERT(it != m_parent->m_sections.end());
-    REALM_ASSERT(m_parent->is_valid() && idx < it->second.indices.size());
+    auto s = it->second.indices.size();
+    if (idx >= s)
+        std::out_of_range(util::format("Requested index %1 greater than max %2", idx, s - 1));
     return m_parent->m_results.get_any(it->second.indices[idx]);
 }
 
 Mixed ResultsSection::key()
 {
-    REALM_ASSERT(is_valid());
+    if (!is_valid())
+        throw std::logic_error("This ResultsSection is not in a valid state.");
     return m_key;
 }
 
 size_t ResultsSection::index()
 {
     util::CheckedUniqueLock lock(m_parent->m_mutex);
+    if (!m_parent->is_valid())
+        throw std::logic_error("The parent Results collection is not valid.");
     m_parent->calculate_sections_if_required();
     auto it = m_parent->m_sections.find(m_key);
-    REALM_ASSERT(m_parent->is_valid() && it != m_parent->m_sections.end());
+    REALM_ASSERT(it != m_parent->m_sections.end());
```

Comment:
This being an assertion means that this function has a precondition of calling `is_valid()` immediately before it as that's the only way to ensure that this assertion never fails, but if that's the case then the first half of the function is redundant.